ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -0,0 +1,205 @@
+	rayDashboardClient.InitClient(clientURL)
+
+	if rayJobInstance.Status.JobId == "" {
+		// Submit the job if no id set
+		jobId, err := rayDashboardClient.SubmitJob(rayJobInstance)
+		if err != nil {
+			r.Log.Error(err, "failed to submit job")
+			err = r.updateState(ctx, rayJobInstance, rayv1alpha1.JobDeploymentStatus_FailedJobDeploy, err)
+			return ctrl.Result{}, err
+		}
+		log.Info("Job succesfully submmited", "jobId", jobId)
+		rayJobInstance.Status.JobId = jobId
+		rayJobInstance.Status.JobStatus = rayv1alpha1.JobStatus_Pending
+		err = r.updateState(ctx, rayJobInstance, rayv1alpha1.JobDeploymentStatus_Running, nil)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, nil
+	}
+
+	// If jobId set already, just check job status
+	jobInfo, err := rayDashboardClient.GetJobInfo(rayJobInstance.Status.JobId)
+	if err != nil {
+		err = r.updateState(ctx, rayJobInstance, rayv1alpha1.JobDeploymentStatus_FailedToGetJobStatus, err)
+		return ctrl.Result{}, err
+	}
+	if jobInfo.JobStatus != rayJobInstance.Status.JobStatus {
+		rayJobInstance.Status.JobStatus = jobInfo.JobStatus
+		err = r.updateState(ctx, rayJobInstance, rayv1alpha1.JobDeploymentStatus_Running, nil)
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *RayJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&rayv1alpha1.RayJob{}).
+		Owns(&rayv1alpha1.RayCluster{}).
+		Owns(&corev1.Service{}).
+		Complete(r)
+}
+
+func (r *RayJobReconciler) getRayJobInstance(ctx context.Context, request ctrl.Request) (*rayv1alpha1.RayJob, error) {
+	rayJobInstance := &rayv1alpha1.RayJob{}
+	if err := r.Get(ctx, request.NamespacedName, rayJobInstance); err != nil {
+		if errors.IsNotFound(err) {
+			rayServiceLog.Info("Read request instance not found error!")
+		} else {
+			rayServiceLog.Error(err, "Read request instance error!")
+		}
+		// Error reading the object - requeue the request.
+		return nil, err
+	}
+	return rayJobInstance, nil
+}
+
+func (r *RayJobReconciler) updateState(ctx context.Context, rayJob *rayv1alpha1.RayJob, status rayv1alpha1.JobDeploymentStatus, err error) error {
+	rayJob.Status.JobDeploymentStatus = status
+	if errStatus := r.Status().Update(ctx, rayJob); errStatus != nil {
+		return fmtErrors.Errorf("combined error: %v %v", err, errStatus)
+	}
+	return err
+}
+
+// TODO: select existing rayclusters by ClusterSelector
+func (r *RayJobReconciler) getOrCreateRayClusterInstance(ctx context.Context, rayJobInstance *rayv1alpha1.RayJob) (*rayv1alpha1.RayCluster, error) {
+	// Update ray cluster
+	var rayClusterInstanceName string
+	if rayJobInstance.Status.RayClusterName != "" {
+		rayClusterInstanceName = rayJobInstance.Status.RayClusterName
+	} else {
+		rayClusterInstanceName = utils.GenerateRayClusterName(rayJobInstance.Name)
```

Comment:
better split saving cluster name into CR state and create Ray Cluster into 2 reconcile loop. Otherwise, if CR state update fails, you will create another RayCluster, cause resource leak.

Reply:
You can check how RayService handles it